src/transformers/modeling_funnel.py

```diff
@@ -0,0 +1,1529 @@
+            token_type_mat = self.stride_pool(token_type_mat, 2)
+            cls_mask = self.stride_pool(cls_mask, 1)
+            attention_mask = self.pool_tensor(attention_mask, mode="min")
+        attention_struct = (position_embeds, token_type_mat, attention_mask, cls_mask)
+        return attention_struct
+
+
+def _relative_shift_gather(x, context_len, shift):
+    batch_size, n_head, seq_len, max_rel_len = x.shape
+    # max_rel_len = 2 * context_len + shift -1 is the numbers of possible relative positions i-j
+
+    # What's next is the same as doing the following gather, which might be clearer code but less efficient.
+    # idxs = context_len + torch.arange(0, context_len).unsqueeze(0) - torch.arange(0, context_len).unsqueeze(1)
+    # # matrix of context_len + i-j
+    # y = x.gather(3, idxs.expand([bs, n_head, context_len, context_len]))
+
+    y = torch.reshape(x, [batch_size, n_head, max_rel_len, seq_len])
+    y = y[:, :, shift:, :]
+    # y = torch.narrow(y, 2, shift, max_rel_len - shift)
+    y = torch.reshape(y, [batch_size, n_head, seq_len, max_rel_len - shift])
+    y = y[..., :context_len]
+    # y = torch.narrow(y, 3, 0, context_len)
+    return y
+
+
+class FunnelRelMultiheadAttention(nn.Module):
+    def __init__(self, config, block_index):
+        super().__init__()
+        self.config = config
+        self.block_index = block_index
+        d_model, n_head, d_head = config.d_model, config.n_head, config.d_head
+
+        self.hidden_dropout = nn.Dropout(config.hidden_dropout)
+        self.attention_dropout = nn.Dropout(config.attention_dropout)
+
+        self.q_head = nn.Linear(d_model, n_head * d_head, bias=False)
+        self.k_head = nn.Linear(d_model, n_head * d_head)
+        self.v_head = nn.Linear(d_model, n_head * d_head)
+
+        self.r_w_bias = nn.Parameter(torch.zeros([n_head, d_head]))
+        self.r_r_bias = nn.Parameter(torch.zeros([n_head, d_head]))
+        self.r_kernel = nn.Parameter(torch.zeros([d_model, n_head, d_head]))
+        self.r_s_bias = nn.Parameter(torch.zeros([n_head, d_head]))
+        self.seg_embed = nn.Parameter(torch.zeros([2, n_head, d_head]))
+
+        self.post_proj = nn.Linear(n_head * d_head, d_model)
+        self.layer_norm = FunnelLayerNorm(d_model, eps=config.layer_norm_eps)
+        self.scale = 1.0 / (d_head ** 0.5)
+
+    def relative_positional_attention(self, position_embeds, q_head, context_len, cls_mask=None):
+        """ Relative attention score for the positional encodings """
+        # q_head has shape batch_size x sea_len x n_head x d_head
+        if self.config.attention_type == "factorized":
+            # Notations from the paper, appending A.2.2, final formula
+            # phi and pi have shape seq_len x d_model, psi and omega have shape context_len x d_model
+            phi, pi, psi, omega = position_embeds
+            # Shape n_head x d_head
+            u = self.r_r_bias * self.scale
```

Comment:
what is `u`?

Reply:
Notations of the paper, appendix A.2.2, final formula :-)

Reply:
link it!